test/index.test.js

```diff
@@ -0,0 +1,58 @@
+const mock = require('mock-fs');
+const MockDate = require('mockdate');
+const {
+  staticUrl,
+  initializeStaticAssetsManager,
+} = require('../src');
+
+const staticAssetVersionsFilename = 'staticAssets.json';
+const staticAssetVersionsFilename2 = 'staticAssets2.json';
+const filename = 'css/test.css';
+const cdnUrl = 'https://cdn.com/css/test.12345.css';
+
+beforeAll(() => {
+  // Creates an in-memory file system
+  mock({
+    [staticAssetVersionsFilename]: JSON.stringify({ [`/public/${filename}`]: cdnUrl }),
+    [staticAssetVersionsFilename2]: JSON.stringify({ }),
+  });
+});
+
+afterAll(() => {
+  mock.restore();
+});
+
+test('should return a correct css cdn url', async () => {
+  await initializeStaticAssetsManager({ staticAssetVersionsFilename });
+  expect(staticUrl(filename)).toBe(cdnUrl);
+});
+
+test('should return local url', async () => {
+  await initializeStaticAssetsManager({
+    staticAssetVersionsFilename: staticAssetVersionsFilename2,
+  });
+
+  expect(staticUrl(filename)).toBe('/public/css/test.css');
+});
+
+test('should initialize a StaticAssetsManager and trigger an error', async () => {
+  try {
+    await initializeStaticAssetsManager({ staticAssetVersionsFilename: 'unknown-file.json' });
+  } catch (err) {
+    expect(err.message).toBe("ENOENT, no such file or directory 'unknown-file.json'");
+  }
+});
+
+test('should return local url', async () => {
+  await initializeStaticAssetsManager({ });
```

Comment:
These tests are really useful to seeing the way this library could be used! I know I created the first version within buffer-marketing from what this is based on, but I'm not sure if my original approach is best for simple consumption in projects. Re-thinking the original approach I took, I think it could be a lot better and potentially have more clearer and simpler variable names and arguments as well as usage for different types of apps. 

Bear with me while I re-think the original code I slapped together 😄 ....I'll try to write an examples of how someone might use this in their code to illustrate my thoughts:

```js
const app = express()
app.get('/', someHandler)

// Version 1
async function start() {
  await initializeStaticAssetsManager({
    staticAssetVersionsFilename: process.env.NODE_ENV === 'production' ? 'staticAssets.json' : null,
  })
  app.listen(8080)
}
start()

// Version 2
initializeStaticAssetsManager({
  staticAssetVersionsFilename: process.env.NODE_ENV === 'production' ? 'staticAssets.json' : null,
}).then(() => app.listen(8080))
```
- We need to use an `async` function or use `then()` to ensure the app doesn't start accepting traffic before the manager is initialized. This could cause errors as `staticUrl()` would call a method on `staticAssetsManager` which would be undefined. Maybe this is OK, but if so, we should document in the readme or find a way to prevent someone from implementing this library without doing so.
- The line checking if the app's in production and sending the standard filename would likely be copied to every single application. I think we may be able to make this automatic, but configurable.

I'm trying to think of a way we could create a simpler way to use this? We also can consider how other projects are using it like here in New Publish which isn't maybe super ideal, but they are loading the file synchronously: https://github.com/bufferapp/buffer-publish/blob/0314d72edf0a11c3bd1df99855e4eb4abcd28044/packages/server/index.js#L39-L56

```js
class StaticAssetManager {
  constructor({
     filename: 'staticAssets.json',
     useLocal: process.env.NODE_ENV === 'development', // default this, but allow for override
     path: '/', // the relative path where static assets are served during local dev
     localPort: null, // useful for teams using webpack dev server running outside the container (ex. publish)
     dir: 'public', // prefix for looking up static files in the staticAssets.json (the local directory prefix)
  }) {
    this.filename = filename
    this.useLocal = useLocal
    this.path = path
    this.localPort = localPort
    this.dir = dir
    this.versions = {}
  }

  load() {
    return new Promise((resolve, reject) => {
      fs.readFile(this.filename, (err, c) => {
        if (err) return reject(new Error(`StaticAssetsManger: Could not load static assets file '${this.filename}'`))
        resolve()
      })
    })
  }
}
let mgnr = null
module.exports.initStaticAssets = (options) => {
  mgnr =  new StaticAssetManager(options)
  if (mgnr.useLocal) return Promise.resolve()
  return mgnr.load()
}
module.exports.initStaticAssetsSync = (options) => {
  mgnr =  new StaticAssetManager(options)
  if (mgnr.useLocal) return
  mgnr.loadSync() // similar to above
}

// Standard usage
initStaticAssets()
  .then(() => app.listen(8080))
// Forcing local files even in prod
initStaticAssets({ useLocal: true }).then(...)
// Alt filename - don't need to check the NODE_ENV
initStaticAssets({ filename: 'assets.json' }).then(...)
// Force using the static assets all the time even in dev (useful for testing)
initStaticAssets({ useLocal: false }).then(...)
// Allow for synchronous usage like in buffer-publish
initStaticAssetsSync({ localPort: 8080 })
// Marketing site
initStaticAssets({ path: '/static', dir: 'public' }).then(...)
```

What do you think?

Reply:
Hey @djfarrelly great suggestions here! It's your turn to bear with me now 😂 
____
> Addition of the sync version

Great addition of the sync version in the package. I was initially thinking of documenting it to make sure people use it with async / await, but it seems I've forgotten myself to use async / await in buffer-login 🤦‍♂️That's a better way to provide both async and sync versions 👍
____
> path: '/', // the relative path where static assets are served during local dev
     localPort: null, // useful for teams using webpack dev server running outside the container (ex. publish)
     dir: 'public', // prefix for looking up static files in the staticAssets.json (the local directory prefix)

I like the renaming of `staticAssetDirLocal` and `staticAssetDirCDN` to  `path` and `dir`, and how they better reflect their role and how to be used.

However, I've found `path` and `dir` (previously, `staticAssetDirLocal` and `staticAssetDirCDN`) quite confusing to use. The following is true in both my and your version, that's regardless of the name :) 

**First use case: Buffer Login SSR w/ React**
The need for those two variables is very tied to this first use case: loading assets from the public folder in React, and how we do things between express locally and our production CDN.

1. Our images assets are all in the `/public` folder, for instance `/public/img/marketing-bg-tl.png`
2. Locally, we don't need to specify `/public` in the assets URL because we told express to retrieve the assets from the `public` folder by default: [here](https://github.com/bufferapp/buffer-login/blob/a05c34cb2f58d04990953a5fe779afa703eb7130/src/app.js#L68) and [here](https://github.com/bufferapp/buffer-login/blob/a05c34cb2f58d04990953a5fe779afa703eb7130/src/routes/login/views/components/MarketingCTA.jsx#L12)
3. When running pre-build.sh, the assets uploader will generate keys using the full path, so keys will look like `public/img/marketing-bg-tl.png => CDN URL`.
4. Then, in production, we need to append `public` to generate the right lookup key: that's where we need the `dir` variable.

**Second use case: Webpack assets**
If we want to make this library useful also for loading webpack assets, then `dir` we need to initialize the lib with `dir: '/'`. The reasons are:

1. Webpack build will generate `vendor.js` file in the main directory `/` not in `/public`
2. This current version will lookup for `public/vendor.js`.
3. When used with webpack dev server, we need to specify not only the port but the full URL `https://local.buffer.com:8080/static/` to be used in development mode: `getStaticURL('vendor.js')` would return `https://local.buffer.com:8080/static/vendor.js`, and [no need to create a local assets object](https://github.com/bufferapp/buffer-account/blob/4312d1e0f331312beb1480accd7641dfd152d071/packages/server/index.js#L22).

This would work for now, while I still find the `dir` approach confusing.

**Third use case: Both at once**
Correct me if I'm wrong, but ultimately it seems we will need both at once. For instance, in buffer-login we will want to load the webpack builds and also load assets from the CDN in React. I think we will hit a wall with the configuration of `path`, `localPort/URL` and `dir` as outlined in the before, they will require different values.

- We may need to rethink a bit how we generate static URL from the lib using those 3 variables.
- At first, to move forward faster, we could provide two distinct functions: one for loading assets (images) in React, and the other for loading build files at the start of the app.
- We could also move the webpack generated files to the `/public` folder before they are being uploaded in the `pre-build.sh` file.

**Other thought**
Finally, this all work well on the server side, but if we want to use this library on the client side, we will need to think of where to keep the data (the list of key/value assets): will it be a store?

Actually do we have the intention of using it on the client side as well? I feel that would potentially be valuable since engineers wouldn't have to care anymore about uploading the assets themselves to the CDN.

___
If you are still there, what do you think of all the above? 😅 😄

Reply:
You've got some great points @philippemiguet. 

**re: awkward path/dir file paths**:
- Maybe we can remove `/public` in both the key and CDN file path? If we wanted this we'd have to ship a breaking change to `buffer-static-upload` which may be a great idea!
- I think that might help remove the `dir` part of the config completely and remove the awkwardness you're sharing 😄 

**re: client side**
We'll need a client side usage of this, but I see that mostly being for images since they'd be used in the React views. Since images don't have versioned/hashed file names we may just need the base directory. Here's a draft on how that could work:
```js
// @bufferapp/static-assets/client.js
let staticFileBaseURL = '/static/' // default for local dev typically via express-static
export setBaseURL = (baseURL) => staticFileBaseURL = baseURL
export staticURL = (file) => `${staticFileBaseURL}${file}`
```
And in the application:
```js
// app.js
import MyApp from './components/MyApp'
import { setBaseURL } from '@bufferapp/static-assets/client'

// this could be set by a global bootstrapped on the page
setBaseURL(window.staticAssetBaseURL) 
// -> we could even make this the default!

React.render(<MyApp/>, document.getElementById('#app'))

// ./components/MyApp.jsx
import { staticURL } from '@bufferapp/static-assets/client'
```
```html
<div id="body></div>
<!-- in dev -->
<script> window.staticAssetBaseURL = '/static/' </script>
<!-- in prod -->
<script> window.staticAssetBaseURL = 'https://static.buffer.com/login/' </script>
```

We could even take this a step further and make some of this more automatic with a piece of express middleware that injects the `staticAssetBaseURL` into `res.locals` but that may be a bit overkill for now 😁

Reply:
I've decided to track any changes w/ the public dir prefix here: https://github.com/bufferapp/buffer-static-upload/issues/15 cc @philippemiguet

Reply:
Hey @djfarrelly, great idea removing the public dir prefix from the generated keys, it will solve the issue.  I left a comment directly in the issue you have opened!

**re: local host and webpack dev server**
I have opened this issue here: . The next version should remove the need for engineers to have to declare their own local assets to mock what the data would look like in production - example [here](https://github.com/bufferapp/buffer-publish/blob/109bbd9f6180f3509922f2485e9b8fee2c9a9030/packages/server/index.js#L42-L59) from Publish repo.

With your idea of passing a `localHostURL` parameters, and `useLocal`, when using `getStatic('vendor.js')`, the lib would generate `https://local.buffer.com:8080/static/vendor.js` right away. 

I wonder if webpack dev server can serve images as well. Curious what will happen when called for images then. Would `https://local.buffer.com:8080/static/buffer.png` work?

Reply:
> I wonder if webpack dev server can serve images as well. Curious what will happen when called for images then. Would https://local.buffer.com:8080/static/buffer.png work?

Good question - I think if webpack dev server is setup to be a proxy, it will serve the image on the express site.

For the other ones, I have added comments in the new issues :D Thanks for tracking all of these things in Github!

Reply:
> Good question - I think if webpack dev server is setup to be a proxy, it will serve the image on the express site.

I'm adding a not to the #3 issue since it's related to passing a localHost URL to support webpack dev server, to figure out the config to make it work altogether.